fsthttp/request.go

```diff
@@ -399,40 +423,308 @@ func (req *Request) Send(ctx context.Context, backend string) (*Response, error)
+	// use guest cache
+
+	if ok, err := fastly.HTTPCacheIsRequestCacheable(req.abi.req); err != nil {
+		return nil, fmt.Errorf("request not cacheable: %v", err)
+	} else if !ok {
+		// no error during lookup but request not cacheable;
+		abiResp, abiBody, err := req.sendWithoutCaching(backend)
+		if err != nil {
+			return nil, err
+
+		}
+
+		resp, err := newResponse(req, backend, abiResp, abiBody)
+		if err != nil {
+			return nil, fmt.Errorf("construct response: %w", err)
+		}
+
+		resp.updateFastlyCacheHeaders(req)
+		return resp, nil
+	}
+
+	var options fastly.HTTPCacheLookupOptions
+	if key := req.CacheOptions.OverrideKey; key != "" {
+		options.OverrideKey(key)
+		req.CacheOptions.OverrideKey = ""
+	}
+
+	// force the lookup to await in the host, retrieving any errors synchronously
+	cacheHandle, err := fastly.HTTPCacheTransactionLookup(req.abi.req, &options)
+	if err != nil {
+		return nil, fmt.Errorf("cache transaction lookup: %w", err)
+	}
+	// in a function so we can change cacheHandle later and have it reflected here
+	defer func() {
+		if cacheHandle != nil {
+			fastly.HTTPCacheTransactionClose(cacheHandle)
 		}
 	}()
+	if err := httpCacheWait(cacheHandle); err != nil {
+		return nil, err
+	}
+
+	// is there a "usable" cached response (i.e. fresh or within SWR period)
+	resp, err := httpCacheGetFoundResponse(cacheHandle, req, backend, true)
 	if err != nil {
-		return nil, fmt.Errorf("poll: %w", err)
+		return nil, err
 	}
 
-	resp, err := newResponse(req, backend, abiResp, abiRespBody)
+	if resp != nil {
+		// got a response from the cache
+
+		// if this is during SWR, we may be the "lucky winner" who is
+		// tasked with performing a background revalidation
+		if ok, _ := httpCacheMustInsertOrUpdate(cacheHandle); ok {
+			pending, err := req.sendAsyncForCaching(ctx, cacheHandle, backend)
+			if err != nil {
+				return nil, err
+			}
+
+			// Wait for the pending respond, then call any after-end hooks
+			go func() {
+				candidate, err := newCandidateFromPendingBackendCaching(pending)
+				if err != nil {
+					// nowhere to log error
+					return
+				}
+				candidate.applyInBackground()
+				fastly.HTTPCacheTransactionClose(cacheHandle)
+			}()
+			// let cache handle be closed in goroutine
+			cacheHandle = nil
+		}
+
+		// Meanwhile, whether fresh or in SWR, we can immediately return
+		// the cached response:
+		resp.updateFastlyCacheHeaders(req)
+		return resp, nil
+	}
+
+	// no cached response
+
+	if ok, _ := httpCacheMustInsertOrUpdate(cacheHandle); ok {
+
+		pending, err := req.sendAsyncForCaching(ctx, cacheHandle, backend)
+		if err != nil {
+			return nil, err
+		}
+
+		candidateResp, err := newCandidateFromPendingBackendCaching(pending)
+		if err != nil {
+			return nil, err
+		}
+
+		resp, err := candidateResp.applyAndStreamBack(req)
+		if err != nil {
+			return nil, err
+		}
+		resp.updateFastlyCacheHeaders(req)
+
+		cacheHandle = nil
```

Comment:
These uses of `cacheHandle` aren't a race?

Reply:
Yes.  Fixed in 678daa0 by passing removing variables that were closed over (`pending` and `cacheHandle`).